### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by skp96**
[Link to story](https://trello.com/c/34LGkFp4/90-error-handling-4)

Highlights:

- Created an `InternalServerError` class that implements an `HttpServerError` interface
- Tested the `InternalServerError` class
- Moved attributes of `Action`s that rely on utility classes, and rewrote them as variables within the `act()` method of `Action`s in order to catch exceptions during `Server` runtime. This ensures the `Server` continues to run and sends an internal server error to the client in case the server code has errors
- Added the `InternalServerError` to the catch block found in the `Server` class

### CODE REVIEW

File: src/main/kotlin/Actions/MethodNotAllowedAction.kt [hunk 1]
```diff
@@ -19,12 +18,12 @@ class MethodNotAllowedAction(private val allowedMethods: MutableSet<String>) : A
...
 
-    private fun generateAllowHeader() {
+    private fun generateAllowHeader(): Map<String, String> {
         val headerValues = mutableListOf<String>()
 
         for (methods in allowedMethods) {
             headerValues += methods
         }
-        headers = mapOf("Allow" to headerValues.joinToString())
+        return mapOf("Allow" to headerValues.joinToString())
```
Comment (k3ntako) [inline src/main/kotlin/Actions/MethodNotAllowedAction.kt:27]: nice! pure functions are great!

File: src/main/kotlin/Server.kt [hunk 2]
```diff
@@ -31,6 +33,8 @@ class Server(private val serverSocket: ServerSocket,
                 writeResponse(response)
             } catch (e: Exception) {
                 println("Something went wrong: ${e.message}")
+                val response = serverError.handleError(responseBuilder)
```
Comment (k3ntako) [inline src/main/kotlin/Server.kt:36]: since you have the error message, do you think that should be included in the response? Or do you think exposing unpredictable error messages a security vulnerability?

**Review by k3ntako — APPROVED** [review]
Nice work!

File: src/main/kotlin/Server.kt (hunk 2 above)
Comment (skp96) [reply src/main/kotlin/Server.kt:36]: My decision to not include the error message in the response wasn't motivated by a security point of view, but more along the lines of the client not needing to know about the internal workings of the server. Additionally, the client can't do much with that information because it is out of their control. However, thank you for bringing security vulnerability up. I did some reading and if the client is a malicious user then they could use this information to attack the service. I don't think providing just the exception message would be enough to expose the service to such a vulnerability. Including a stack trace; however, could provide the attacker with more information.

File: src/test/kotlin/HttpResponseBuilderTest.kt [hunk 3]
```diff
@@ -3,6 +3,7 @@ import kotlin.test.*
 import response.HttpResponseBuilder
 import httpstatus.HttpStatus
 import org.json.JSONObject
+import kotlin.math.exp
```
Comment (cdemyanovich) [inline src/test/kotlin/HttpResponseBuilderTest.kt]: Why is this import added?

Comment (skp96) [reply src/test/kotlin/HttpResponseBuilderTest.kt]: I believe it was auto imported. I have to figure out how to turn off the auto-import feature in Intellij

**Review by cdemyanovich — APPROVED** [review]
